runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java

```diff
@@ -394,6 +395,15 @@ public void stop(boolean shutdown) {
         stop();
     }
 
+    /**
+     * {@inheritDoc
+     */
+    @Override
+    protected void finalize() throws Throwable {
```

Comment:
don't use finalize, it's a bad practice

Reply:
https://howtodoinjava.com/java/basics/why-not-to-use-finalize-method-in-java/

Reply:
I'm fully aware of these caveats when I create this path. But this is the only approach for now to solve this race condition and meanwhile clean up the resources (i.e. channels) without relying on client invoking it.

The biggest argument in this link is that "finalize() execution is not guaranteed", which I already thought about. However this doesn't matter, even if it's not executed, the result is just some channels remain opening. Still it's better than not providing finalize(). 

My point is, providing finalize() is giving it another chance to clean up the resource, even it's not executed, just leave it. I'm willing to discuss any better ideas if you have, otherwise I think finalize() is ok to be used here.

Reply:
I don't see any reasons to use finalize, I also don't see how finalize() can sole the "race condition" issue

Reply:
Can you please find a better way, or/and maybe redesign this code, to solve the issue without finalize

Reply:
If you want to know the reason to use finalize to solve the race condition, I can explain to you offline, I've already spent some time thinking about better approaches before I go with finalize, but looks like this is the best and cheapest way. Maybe you will have a better idea after we discuss.